pkg/gardenlet/bootstrap/certificate/certificate_rotation_test.go

```diff
@@ -0,0 +1,316 @@
+)
+
+var _ = Describe("Certificates", func() {
+	var (
+		log logrus.FieldLogger = logger.NewFieldLogger(logger.NewLogger("info"), "test", "certificate-test")
+
+		ctrl            *gomock.Controller
+		k8sGardenClient *mock.MockInterface
+		k8sClient       client.Client
+
+		seedName = "test"
+		seed     *gardencorev1beta1.Seed
+
+		gardenClientConnection = &config.GardenClientConnection{
+			KubeconfigSecret: &corev1.SecretReference{
+				Name:      "gardenlet-kubeconfig",
+				Namespace: "garden",
+			},
+		}
+
+		approvedCSR = certificatesv1beta1.CertificateSigningRequest{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "watched-csr",
+			},
+			Status: certificatesv1beta1.CertificateSigningRequestStatus{
+				Conditions: []certificatesv1beta1.CertificateSigningRequestCondition{
+					{
+						Type: certificatesv1beta1.CertificateApproved,
+					},
+				},
+				Certificate: []byte("my-cert"),
+			},
+		}
+
+		deniedCSR = certificatesv1beta1.CertificateSigningRequest{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "watched-csr",
+			},
+			Status: certificatesv1beta1.CertificateSigningRequestStatus{
+				Conditions: []certificatesv1beta1.CertificateSigningRequestCondition{
+					{
+						Type: certificatesv1beta1.CertificateDenied,
+					},
+				},
+			},
+		}
+	)
+
+	BeforeEach(func() {
+		ctrl = gomock.NewController(GinkgoT())
+		k8sGardenClient = mock.NewMockInterface(ctrl)
+
+		seed = &gardencorev1beta1.Seed{
+			ObjectMeta: metav1.ObjectMeta{Name: seedName},
+		}
+
+		// create one fake client to both talk to the Seed API server and Garden API server
+		s := runtime.NewScheme()
+
+		// required schemes
+		Expect(corev1.AddToScheme(s)).ToNot(HaveOccurred())
+		Expect(gardencorescheme.AddToScheme(s)).ToNot(HaveOccurred())
+
+		k8sClient = fakeclient.NewFakeClientWithScheme(s, seed)
+		k8sGardenClient.EXPECT().Client().Return(k8sClient).AnyTimes()
+	})
+
+	AfterEach(func() {
+		ctrl.Finish()
+	})
+
+	Describe("#rotateCertificate", func() {
+		It("should not return an error", func() {
+			kubeClient := fake.NewSimpleClientset(&approvedCSR)
+			k8sGardenClient.EXPECT().Kubernetes().Return(kubeClient)
+
+			// mock the update of the secret in the seed with the rotated kubeconfig
+			certClientConfig := &rest.Config{Host: "testhost", TLSClientConfig: rest.TLSClientConfig{
+				Insecure: false,
+				CAFile:   "filepath",
+			}}
+
+			k8sGardenClient.EXPECT().RESTConfig().Return(certClientConfig)
+			k8sGardenClient.EXPECT().DirectClient().Return(k8sClient)
+
+			fakeClientMap := fakeclientmap.NewClientMap().
+				AddClient(keys.ForGarden(), k8sGardenClient).
+				AddClient(keys.ForSeed(seed), k8sGardenClient)
+
+			err := rotateCertificate(context.TODO(), log, gardenClientConnection, fakeClientMap, seedName)
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		It("should return an error - CSR is denied", func() {
+			kubeClient := fake.NewSimpleClientset(&deniedCSR)
+			k8sGardenClient.EXPECT().Kubernetes().Return(kubeClient)
```

Comment:
You don't have to mock `pkg/client/kubernetes.Interface`, probably using `pkg/client/kubernetes/fake.ClientSet` is easier ;)
E.g.:
```go
k8sGardenClient := fakeclientset.NewClientSetBuilder().WithKubernetes(kubeClient).Build()
```